payments/osp-handler/src/main/java/cm/aptoide/pt/osp_handler/handler/OSPHandler.kt

```diff
@@ -0,0 +1,49 @@
+package cm.aptoide.pt.osp_handler.handler
+
+import android.net.Uri
+import cm.aptoide.pt.osp_handler.di.DefaultOemIdPackageName
+import cm.aptoide.pt.osp_handler.handler.OSPUriConstants.Parameters
+import cm.aptoide.pt.osp_handler.handler.exception.MissingDataParseException
+import cm.aptoide.pt.payment_manager.manager.domain.PurchaseRequest
+import javax.inject.Inject
+import javax.inject.Singleton
+
+@Singleton
+class OSPHandlerImpl @Inject constructor(
+  private val oemIdExtractorService: OemIdExtractorService,
+  @DefaultOemIdPackageName private val oemIdDefaultPackageName: String?,
+) : OSPHandler {
+
+  override suspend fun extract(uri: Uri?): PurchaseRequest? {
+    if (uri == null) return null
+
+    val scheme = uri.scheme
+    val host = uri.host
+    val path = uri.path
+    val parameters = mutableListOf<Pair<String, String>>().also { list ->
+      uri.queryParameterNames.forEach { name ->
+        uri.getQueryParameter(name)?.let { param -> list.add(name to param) }
+      }
+    }
+    val domain = parameters.find { it.first == Parameters.DOMAIN }?.second
+      ?: throw MissingDataParseException("OSP uri must contain the domain name")
+    val oemId = oemIdExtractorService.extractOemId(oemIdDefaultPackageName ?: domain)
+    return PurchaseRequest(
+      scheme = scheme ?: throw MissingDataParseException("OSP uri must contain the scheme"),
+      host = host ?: throw MissingDataParseException("OSP uri must contain the host"),
+      path = path ?: throw MissingDataParseException("OSP uri must contain the path"),
+      product = parameters.find { it.first == Parameters.PRODUCT }?.second,
```

Comment:
If `parameters` was `Map` you would not need to `find` each value iterating over the list every time.